zap/src/output/mod.rs

```diff
@@ -1,40 +1,87 @@
 use std::fmt::Display;
 
-use crate::{
-	irgen::{gen_ser, Expr, Stmt, Var},
-	parser::{Ty, TyDecl},
-	util::NumTy,
-};
+use crate::{irgen::Stmt, parser::Ty, util::NumTy};
 
 pub mod client;
 pub mod server;
 
-impl Display for TyDecl {
-	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-		let name = &self.name;
-		let ty = &self.ty;
-
-		writeln!(f, "export type {name} = {ty}")?;
+pub trait Output {
+	fn push(&mut self, s: &str);
+	fn indent(&mut self);
+	fn dedent(&mut self);
+	fn push_indent(&mut self);
 
-		writeln!(f, "function types.write_{name}(value: {name})")?;
+	fn push_line(&mut self, s: &str) {
+		self.push_indent();
+		self.push(s);
+		self.push("\n");
+	}
 
-		for stmt in gen_ser(ty, "value", false) {
-			writeln!(f, "\t{stmt}")?;
+	fn push_stmt(&mut self, stmt: &Stmt) {
+		if matches!(stmt, Stmt::ElseIf(..) | Stmt::Else | Stmt::End) {
+			self.dedent();
 		}
 
-		writeln!(f, "end;")?;
+		match &stmt {
+			Stmt::Local(name, expr) => {
+				if let Some(expr) = expr {
+					self.push_line(&format!("local {name} = {expr}"));
+				} else {
+					self.push_line(&format!("local {name}"));
+				}
+			}
+
+			Stmt::Assign(var, expr) => self.push_line(&format!("{var} = {expr}")),
+			Stmt::Error(msg) => self.push_line(&format!("error({msg})")),
+			Stmt::Assert(cond, msg) => match msg {
+				Some(msg) => self.push_line(&format!("assert({cond}, {msg})")),
+				None => self.push_line(&format!("assert({cond})")),
+			},
+
+			Stmt::Call(var, method, args) => match method {
+				Some(method) => self.push_line(&format!(
+					"{var}:{method}({})",
```

Comment:
What about callbacks? `::Call` doesn't seem to account for stuff like `var.callback()`...

Reply:
the `Var` enum accounts for indexing separately.